### PR TITLE
Adding CBMC proof for TCP handle state

### PR DESCRIPTION
**Pull request by dlohar**
Description
-----------
This PR includes the CBMC proof for memory safety of the prvTCPHandleState function.

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [ ] My code is Linted.

### CODE REVIEW

File: libraries/freertos_plus/standard/freertos_plus_tcp/source/FreeRTOS_TCP_IP.c [hunk 1]
```diff
@@ -2671,6 +2671,8 @@ BaseType_t xSendLength = 0;
 uint32_t ulReceiveLength;	/* Number of bytes contained in the TCP message. */
 uint8_t *pucRecvData;
 uint32_t ulSequenceNumber = FreeRTOS_ntohl (pxTCPHeader->ulSequenceNumber);
+configASSERT(pxSocket);
```
Comment (danielsn) [inline libraries/freertos_plus/standard/freertos_plus_tcp/source/FreeRTOS_TCP_IP.c]: Would it make sense to move this assert higher, i.e. make it a precondition

File: tools/cbmc/proofs/TCP/prvTCPHandleState/TCPHandleState_harness.c [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+#include "FreeRTOS_IP_Private.h"
+#include "FreeRTOS_TCP_IP.h"
+
+/* This proof assumes that prvTCPPrepareSend and prvTCPReturnPacket are correct.
+These functions are proved to be correct separately. */
+
+
+BaseType_t publicTCPHandleState( FreeRTOS_Socket_t *pxSocket, NetworkBufferDescriptor_t **ppxNetworkBuffer );
+
+/* Implementation of safe malloc */
```
Comment (danielsn) [inline tools/cbmc/proofs/TCP/prvTCPHandleState/TCPHandleState_harness.c]: This is a useful utility function. Make a utils folder and then you can reuse it.

File: tools/cbmc/proofs/TCP/prvTCPHandleState/TCPHandleState_harness.c [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+void *safeMalloc(size_t xWantedSize) {
+	if(xWantedSize == 0) {
```
Comment (danielsn) [inline tools/cbmc/proofs/TCP/prvTCPHandleState/TCPHandleState_harness.c]: in c-common, we bound to never allow malloc of sizes greater than CBMC can handle. Is that a concern here?

File: tools/cbmc/proofs/TCP/prvTCPHandleState/TCPHandleState_harness.c [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+
+void harness() {
+	FreeRTOS_Socket_t *pxSocket = safeMalloc(sizeof(FreeRTOS_Socket_t));
+	if (pxSocket) {
+		pxSocket->u.xTCP.rxStream = safeMalloc(sizeof(StreamBuffer_t));
+		pxSocket->u.xTCP.txStream = safeMalloc(sizeof(StreamBuffer_t));
+		pxSocket->u.xTCP.pxPeerSocket = safeMalloc(sizeof(FreeRTOS_Socket_t));
+		__CPROVER_assume(pxSocket->u.xTCP.xTCPWindow.ucOptionLength == ipSIZE_TCP_OPTIONS/sizeof(uint32_t));
+		__CPROVER_assume(pxSocket->u.xTCP.uxRxWinSize >= 0 && pxSocket->u.xTCP.uxRxWinSize <= 0xfffful);
```
Comment (danielsn) [inline tools/cbmc/proofs/TCP/prvTCPHandleState/TCPHandleState_harness.c]: Magic constant.

File: tools/cbmc/proofs/TCP/prvTCPHandleState/TCPHandleState_harness.c [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+
+void harness() {
+	FreeRTOS_Socket_t *pxSocket = safeMalloc(sizeof(FreeRTOS_Socket_t));
+	if (pxSocket) {
+		pxSocket->u.xTCP.rxStream = safeMalloc(sizeof(StreamBuffer_t));
+		pxSocket->u.xTCP.txStream = safeMalloc(sizeof(StreamBuffer_t));
+		pxSocket->u.xTCP.pxPeerSocket = safeMalloc(sizeof(FreeRTOS_Socket_t));
+		__CPROVER_assume(pxSocket->u.xTCP.xTCPWindow.ucOptionLength == ipSIZE_TCP_OPTIONS/sizeof(uint32_t));
```
Comment (danielsn) [inline tools/cbmc/proofs/TCP/prvTCPHandleState/TCPHandleState_harness.c]: Why this size?

File: tools/cbmc/proofs/TCP/prvTCPHandleState/TCPHandleState_harness.c [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+	if(xWantedSize == 0) {
+		return NULL;
+	}
+	uint8_t byte;
+	return byte ? malloc(xWantedSize) : NULL;
+}
+
+
+void harness() {
+	FreeRTOS_Socket_t *pxSocket = safeMalloc(sizeof(FreeRTOS_Socket_t));
```
Comment (danielsn) [inline tools/cbmc/proofs/TCP/prvTCPHandleState/TCPHandleState_harness.c]: I assume there is other code that will need to take a ```FreeRTOS_Socket_t``` as input.  Making a ```ensure_FreeRTOS_Socket_t_is_allocated()``` and ```ensure_FreeRTOS_Socket_t_is_valid()``` functions allows code reuse.

File: tools/cbmc/proofs/TCP/prvTCPHandleState/TCPHandleState_harness.c [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+		__CPROVER_assume(pxSocket->u.xTCP.uxRxWinSize >= 0 && pxSocket->u.xTCP.uxRxWinSize <= 0xfffful);
+		__CPROVER_assume(pxSocket->u.xTCP.usInitMSS == 2);
+	}
+	
+	NetworkBufferDescriptor_t *pxNetworkBuffer = safeMalloc(sizeof(NetworkBufferDescriptor_t));
+	if(pxNetworkBuffer) {
+		pxNetworkBuffer->pucEthernetBuffer = safeMalloc(sizeof(TCPPacket_t));
+	}
+	if (pxSocket && pxNetworkBuffer && pxNetworkBuffer->pucEthernetBuffer) {
+		publicTCPHandleState(pxSocket, &pxNetworkBuffer);
```
Comment (danielsn) [inline tools/cbmc/proofs/TCP/prvTCPHandleState/TCPHandleState_harness.c]: This returns an error code.  What properties do we expect to be true if error is returned - perhaps that the input was unmodified? What do we expect to be true if error was not returned.  i.e. surely this function has post-conditions? What are they? Can we check them?

File: libraries/freertos_plus/standard/freertos_plus_tcp/source/FreeRTOS_TCP_IP.c (hunk 1 above)
Comment (dlohar) [reply libraries/freertos_plus/standard/freertos_plus_tcp/source/FreeRTOS_TCP_IP.c]: Moved it.

File: tools/cbmc/proofs/TCP/prvTCPHandleState/TCPHandleState_harness.c (hunk 4 above)
Comment (dlohar) [reply tools/cbmc/proofs/TCP/prvTCPHandleState/TCPHandleState_harness.c]: Changed it to size_t.

File: tools/cbmc/proofs/TCP/prvTCPHandleState/TCPHandleState_harness.c (hunk 7 above)
Comment (dlohar) [reply tools/cbmc/proofs/TCP/prvTCPHandleState/TCPHandleState_harness.c]: I have not done these checks before. I discussed with Mark and we decided to proof memory safety first. After we finish it we will revisit these proofs and implement these checks.

File: tools/cbmc/proofs/TCP/prvTCPHandleState/TCPHandleState_harness.c (hunk 3 above)
Comment (dlohar) [reply tools/cbmc/proofs/TCP/prvTCPHandleState/TCPHandleState_harness.c]: Till now we do not have any problem, but we will keep an eye on it.

File: tools/cbmc/proofs/TCP/prvTCPHandleState/TCPHandleState_harness.c (hunk 5 above)
Comment (dlohar) [reply tools/cbmc/proofs/TCP/prvTCPHandleState/TCPHandleState_harness.c]: This is the number of valid bytes in ulOptionsData[]. ulOptionsData[] is defined as uint32_t ulOptionsData[ipSIZE_TCP_OPTIONS/sizeof(uint32_t)]. I think the ucOptionLength should be sizeof(uint32_t) * ipSIZE_TCP_OPTIONS/sizeof(uint32_t). I changed ucOptionLength to that. Do you think it is reasonable?

File: tools/cbmc/proofs/TCP/prvTCPHandleState/TCPHandleState_harness.c (hunk 2 above)
Comment (dlohar) [reply tools/cbmc/proofs/TCP/prvTCPHandleState/TCPHandleState_harness.c]: I have created a folder inside `cbmc` folder and moved `safeMalloc` implementation there.

File: tools/cbmc/proofs/TCP/prvTCPHandleState/TCPHandleState_harness.c (hunk 4 above)
Comment (danielsn) [reply tools/cbmc/proofs/TCP/prvTCPHandleState/TCPHandleState_harness.c]: I'm somewhat confused by ```sizeof(size_t)```.  Did you mean ```SIZE_MAX```?

File: tools/cbmc/proofs/TCP/prvTCPHandleState/TCPHandleState_harness.c (hunk 6 above)
Comment (dlohar) [reply tools/cbmc/proofs/TCP/prvTCPHandleState/TCPHandleState_harness.c]: Added these two functions.

File: tools/cbmc/proofs/TCP/prvTCPHandleState/TCPHandleState_harness.c (hunk 4 above)
Comment (dlohar) [reply tools/cbmc/proofs/TCP/prvTCPHandleState/TCPHandleState_harness.c]: uxRxWinSize is initialized as `size_t`.

**Comment by dlohar** [comment]
@danielsn I have addressed the comments.

File: tools/cbmc/proofs/utility/memory_assignments.c [hunk 8]
```diff
@@ -0,0 +1,25 @@
+#define ensure_memory_is_valid( px ) (px != NULL)
+
+/* Implementation of safe malloc */
+void *safeMalloc(size_t xWantedSize) {
+	if(xWantedSize == 0) {
+		return NULL;
+	}
+	uint8_t byte;
+	return byte ? malloc(xWantedSize) : NULL;
```
Comment (htibosch) [inline tools/cbmc/proofs/utility/memory_assignments.c]: uint8_t byte;
    return byte ? malloc(xWantedSize) : NULL;
	
Maybe I miss something, but 'byte' is an uninitialised variable, used in an expression. Is that was is intended?
Hein

Comment (dlohar) [reply tools/cbmc/proofs/utility/memory_assignments.c]: For our proofs CBMC assumes an unconstrained uint8_t value for byte.

**Pull request closed without merging**